### PR TITLE
Add PyTorch Glossary

### LINKED ISSUE #40640: Need glossary documentation
PyTorch has accumulated a lot of lingo that are fuzzily-defined. Need a glossary table that gives a clear definition to commonly used terms to facilitate discussion and communication. 

cc @jlin27

### PR DESCRIPTION
**Pull request by gmagogsfm**

### CODE REVIEW

File: GLOSSARY.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+operation library on which all else is built.
+
+#### Operation
+A unit of work. For example, a MatMul is an operation.
+
+#### Native Operation
+An operation that comes natively with PyTorch ATen.
+
+#### Custom Operations
+Operations that are defined by users and are usually Compound Operations.
```
Comment (kshitij12345) [inline GLOSSARY.md:17]: Maybe an example would be nice.

File: GLOSSARY.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+#### ATen
+Short for "A Tensor Library". The foundational tensor and mathematical
+operation library on which all else is built.
+
+#### Operation
+A unit of work. For example, a MatMul is an operation.
+
+#### Native Operation
+An operation that comes natively with PyTorch ATen.
```
Comment (kshitij12345) [inline GLOSSARY.md:14]: An example like above, or maybe an actual Python Operation `torch.add` would be nice

File: GLOSSARY.md (hunk 1 above)
Comment (gmagogsfm) [reply GLOSSARY.md]: Done.

File: GLOSSARY.md (hunk 2 above)
Comment (gmagogsfm) [reply GLOSSARY.md]: Done.

File: GLOSSARY.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+#### Operation
+A unit of work. For example, the work of matrix multiplication is an operation
+called aten::matmul.
+
+#### Native Operation
+An operation that comes natively with PyTorch ATen, for example aten::matmul.
+
+#### Custom Operation
+An Operations that is defined by users and are usually Compound Operations.
```
Comment (SplitInfinity) [inline GLOSSARY.md:18]:
```suggestion
An Operation that is defined by users and is usually a Compound Operation.
```

File: GLOSSARY.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+#### Kernel
+Implementation of a PyTorch operation, specifying what should be done when an 
+operation executes.
+
+#### Compound Operation
+A Compound Operation is composed of other operations. Its kernel is usually
+device-agnostic. Normally it doesn't have its own derivative functions defined.
+Instead, AutoGrad automatically computes its derivative based on operations it
+used.
```
Comment (SplitInfinity) [inline GLOSSARY.md:31]:
```suggestion
uses.
```

**Review by SplitInfinity — COMMENTED** [review]
It seems to me that these terms pertain to a specific part of PyTorch (that I personally haven't had to touch). Maybe we should break this down into sections? E.g. core, distributed, JIT, etc.

In addition, I wonder if there are better tools or formats to use for this. I guess people can Ctrl + F but I think everyone expects glossaries to be sorted alphabetically, and that is something that will be tedious to do by hand.

File: GLOSSARY.md (hunk 3 above)
Comment (gmagogsfm) [reply GLOSSARY.md]: Done.

File: GLOSSARY.md (hunk 4 above)
Comment (gmagogsfm) [reply GLOSSARY.md]: Done.

**Comment by gmagogsfm** [comment]
> It seems to me that these terms pertain to a specific part of PyTorch (that I personally haven't had to touch). Maybe we should break this down into sections? E.g. core, distributed, JIT, etc.
> 
> In addition, I wonder if there are better tools or formats to use for this. I guess people can Ctrl + F but I think everyone expects glossaries to be sorted alphabetically, and that is something that will be tedious to do by hand.

Good point. I am not aware of better tools to use for this, maybe github wiki? But then it would not be very visible for fb internal.

Anyway, I added a table of content, vscode actually has an extension that can automatically update table of content upon file save, so that should help with indexing and alleviate the burden of maintenance.

**Comment by gmagogsfm** [comment]
Gentle ping.

**Review by suo — APPROVED** [review]
lgtm!

**Comment by suo** [comment]
on the "should this be a wiki" issue—I think we can just commit in the repo for now and see what happens

**Pull request closed without merging**